Replace broken mesh smoothing with Laplacian

kronbichler opened the pull request:
While working with the improved mapping - hoping that everything would be robust now - I discovered a very fundamental issue that affects both the grid refinement algorithms as well as `MappingQGeneric::compute_mapping_support_points` - thus the *epic* label. We always use the Laplace smoothing to compute new points from surrounding entities. In mesh refinement for instance, we let the new point inside a quad be determined not only by the four vertices but also by the four new points on the lines. 

**Short version: I believe that the current weights are not suitable for what we are doing and we should change them before the release.**

The intuitive reason why the current approach is wrong is the following: Suppose we are given  vertices and additionally points in midpoints of the lines. When we want to compute a new point in the center of the cell, we do not really want to take an *average* of the eight surrounding points, but rather *interpolate* from the circumference into the interior. This is the functional principle of the transfinite interpolation manifold that we have added in #4404. The same mechanism should be applied when a single cell gets refined and we already know the mid points of the lines. To switch to a better scheme, we thus only need to choose different weights on the points of the faces versus the vertices of the cell.

Let me show the difference this makes. This is a picture reproducing of what we explain in http://dealii.org/developer/doxygen/deal.II/group__manifold.html when creating a shell mesh with 3 cells:

![grid_bad](https://user-images.githubusercontent.com/8276200/32383862-73e92ede-c0b9-11e7-912c-b79fcee5ff0b.png)

What happens for this mesh is that at the first refinement (with 3 cells around the rim) we place the new middle points on the lines according to the circle. When we compute the new point in the center of the 3 cells an average actually ends up *outside* the circle. But we do not want to take an average here, we want to interpolate between the two points on the lines. This is done by replacing the weights `{1./16, 1./16, 1./16, 1./16, 3./16, 3./16, 3./16, 3./16}` by `{-1./4, -1./4, -1./4, -1./4, 1./2, 1./2, 1./2, 1./2}`. It looks weird with negative weights at first sight but the mechanism does what we want: In case of a cell along the rim, the negative weights of `-1./4` in the vertices balance with the center of the two straight lines in radial direction that get weight `1./2`. Thus, we only have the two center points in circumferential direction left whose average we're taking. I've tested this on many cases and always works. Here is the mesh (starting from the same weird coarse mesh), with only setting a boundary descirption and no volume description (which would of course have fixed the mesh):

![grid_good](https://user-images.githubusercontent.com/8276200/32384184-5658da08-c0ba-11e7-98dd-9072f3ada9f9.png)

Of course, after having refined the 3 cells once, we have faces in circumferential direction that are straight, such that subsequent refinement does not place the new points on a circular shape any more. So there are no miracles, it's just a better way of taking the surrounding into account.

We also get much better meshes in 3d - this is a cut through the center of a 3D ball mesh in the standard code, refined once globally (into 56 cells) and then outputted on a `MappingQGeneric` with 10 sub-patches in `DataOut::build_patches`:

![ball_manifold_bad](https://user-images.githubusercontent.com/8276200/32384260-9c48b722-c0ba-11e7-8b69-4acd7be14524.png)

and this with my modification:

![ball_manifold_good](https://user-images.githubusercontent.com/8276200/32384281-b2c6548c-c0ba-11e7-900a-a3cd01d7d411.png)

Again, this is as good as it can get from a boundary description alone. Of course, everyone on a `TransfiniteInterpolationManifold` did already have this (and better) behavior, but I assume the default user did not and this would give appreciably better meshes.

The trouble we had with *averaging* rather than *interpolating* does not stop here: One of the motivations of `TransfiniteInterpolationManifold` that was written by @luca-heltai and me was to fix convergence rates at high order basis functions. We believed (see also #3874) that this was intrinsic to switching from a curved description to a straight description inside a single layer. But my initial tests with the proposed code show that we do restore the optimal convergence rates when going from the bad Laplace smoothing to a proper interpolation with the concepts of transfinite interpolation also *inside a single layer of curved cells at the boundary*. So, it seems that I (and possibly also @drwells) have been chasing the wrong rabbit for a while. (This is so fundamental that I spent all of today on thoroughly understanding what is going on.) The problem with the Laplace smoothing is that the *averaging* has no notion of the shape drawn by the lines around the vertices, essentially assuming an isotropic non-deformed setup. When used inside `MappingQGeneric`, this "naive" smoothing then puts points in positions that make the mapped function space miss some low order polynomials that are required for convergence to a certain order: Essentially, it does so starting from cubic polynomials onwards because its solution is harmonic in the second derivative. And a missing cubic contribution means convergence of order 3 on the affected cells, or 3.5 when looking at global L2 errors. Bingo. Once we interpolate the point positions from the position of the boundary points, taking into account the curved boundary, we obviously preserve the properties of the polynomial space - transfinite interpolation is, after all, nothing else than a linear interpolation between the edges' curved shapes with negative compensation for taking the vertices twice.

This PR makes the basic fixes in the triangulation and the Q mapping. There will certainly be a number of tests affected because this is a fundamental change. It will likely also affect a large number of user codes. I believe that the upcoming major release is the right point to make this break, but before going on with the boring part I want to hear opinions.

The smoothing algorithm appears to have been written by Ralf Hartmann who has left the project a long time ago. So you @bangerth might be the only one left with knowledge of the Laplace smoothing and the mapping report https://www.dealii.org/reports/mapping_q/index.html where the concepts of mesh smoothing are explained. Do you have objections? In particular, can somebody imagine cases where the proposed method will fail (e.g. due to negative weights)? I argue that it will be robust on all the tasks we perform: We interpolate from vertices, lines (and quads in 3D) into the interior. Each higher hierarchy has been obtained either from the manifold or from an interpolation between the known good states of the lower level entities. No way this could go wrong.

tjhei commented:
> And a missing cubic contribution means convergence of order 3 on the affected cells, or 3.5 when looking at global L2 errors. Bingo.

Whoa! I have been discussing this with my students, @drwells,  and @bangerth after seeing the same suboptimal convergence order with higher order mappings.

I have no understanding of the mappingq internals but your pictures are definitely convincing. Do you think it makes sense to search for higher order mappings for FEM in the literature?

I can try out if this fixes the convergence rates for me.

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -1202,8 +1202,8 @@ namespace Manifolds
-            std::fill(points_weights.second.begin(), points_weights.second.begin()+4, 1.0/16.0);
-            std::fill(points_weights.second.begin()+4, points_weights.second.end(), 3.0/16.0);
+            std::fill(points_weights.second.begin(), points_weights.second.begin()+4, -0.25);
+            std::fill(points_weights.second.begin()+4, points_weights.second.end(), 0.5);
```
It's now no longer a Laplace smoothing, so the `with_laplace` flag is not the correct name any more. Can you update this in the documentation of the function?

kronbichler replied:
I will definitely change this, part of what I considered "boring" in my original post.

masterleinad commented:
/run-tests

kronbichler commented:
> I think it would be nice if you could morph your description of this PR -- including the pictures, and maybe the first couple of history from above about how one can intuitively understand that the edge midpoint weights need to be bigger

The question is of course where we want to add this description. I was first thinking about the manifold description where I got this from, http://dealii.org/developer/doxygen/deal.II/group__manifold.html. While it's not wrong to put there because the proposed change cannot do miracles, I think the weights for the triangulation would deserve a description there. The mapping should adapt what we wrote recently in `MappingQGeneric` regarding the weights. I'm not sure if I'm finding all spots where this needs modification, so I appreciate if you (and other people looking at this PR) have an eye on the documentation in the next weeks to find other spots that will get missed with this PR.

(And my apologies for mis-representing the history here - I just assumed from the mapping page https://www.dealii.org/reports/mapping_q/index.html that has Ralf Hartmann on it that this must be the main author. Anyway, that page obviously needs to be updated as well...)

drwells commented:
This is really good work. I'll spend a few hours really reading into this today.

bangerth commented:
Yes, I don't really know either where to put this piece of documentation. Is there anything in the class documentation of `Triangulation` that talks about how a cell is refined? Otherwise, somewhere in the documentation of either the `Mapping` class or the mappings group page.

As for history: You didn't mis-represent much. The mappings were written by Ralf and Guido. It's just that the Laplace weights for finding the center point predates the mappings.

drwells commented:
AFAIK the 'reports' are not accessible from the main manual. I found them only by googling. I think that we should just write another module on higher-order mappings: I assumed that those pages were the successor to the reprots.

kronbichler commented:
For reference: The patch causes around 190 test failures that need to get the output updated:
https://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=12845
(there are around 20 failing tests due to using FMA on my machine).

I need to check whether there are no "real" failures among them.

kronbichler commented:
>  I think that we should just write another module on higher-order mappings: I assumed that those pages were the successor to the reprots.

I agree. When I find the time I will copy the discussion above to such a file. Then @drwells or anyone else can fill in more info as we gain more experience.

davydden commented:
> I think it would be nice if you could morph your description of this PR -- including the pictures, and maybe the first couple of history from above about how one can intuitively understand that the edge midpoint weights need to be bigger -- into a section in the documentation of one of the involved classes.

👍

drwells commented:
I can confirm that this fixes the issues with the 'bad laplace' solver I wrote to demonstrate the issue (i.e., the test case I have discussed with @tjhei and @bangerth).

I asked the finite differencers here about this issue [1] earlier this fall. They unanimously agreed that, with finite differences and overlapping grids, one should be able to get arbitrarily high-order convergence to a Laplace equation solved on a disk with a Cartesian background grid and a boundary-fitted annular grid that only has a fixed depth (i.e., the boundary is approximated by an annulus (which partially overlaps the cartesian grid covering most of the geometry) with calculations done in polar coordinates which only has four points in the radial direction regardless of the total number of points). Put another way: it seems to me that, as long as the grid is 'nice', that we should be able to use a bilinear map everywhere but the boundary cells (i.e., `MappingQ` is OK and we don't need to use isoparametric elements on the interior).

Here is what I now understand the problem to be (please correct me if I am wrong): by solving a Laplace equation for the mapping support points and weights we, in some circumstances, picked a basis for the mapping that was not isoparametric (i.e., we were missing cubics). Subparametric finite elements don't have the same guarantees so we lose orders of convergence locally. Is this correct?

As always, @kronbichler, I am in awe of your investigative skills. A question: where did you get the new weights from?

[1] As I have mentioned in the past in conversations, there is a [large finite difference group](http://www.overtureframework.org/) here at RPI that spends a lot of time worrying about exact geometry representation with overlapping grids.

tjhei commented:
This also fixes #5039

kronbichler commented:
> A question: where did you get the new weights from?

The weights are what you get when you apply some kind of transfinite interpolation for the given points in the surrounding - that's at least the weights you would get there and that lead me into this direction. But in that case, transfinite interpolation is not the right expression. It really is just a linear interpolation between two pairs of the midpoints of opposite lines into the center (weights 0.5 each), and then subtract 0.25 times the four vertex positions to end up with a sum of 1 for the weights. You can imagine the mechanism for a face with two opposite faces that are curved, say the inner and outer rim of the shell in my first picture, and the other two faces straight-sided. Then you want to take the average of the two curved faces, disregarding the contribution of the two straight lines. And that is indeed what happens, because the midpoint of the faces is exactly 0.5 times the sum of the two surrounding vertices.

kronbichler commented:
> Do you think it makes sense to search for higher order mappings for FEM in the literature?

I had skimmed through the book "Higher-Order Finite Element Methods" by Solin, Segeth, Dolezel and they advocate for using transfinite interpolation. But I have mostly picked up statements like "do this, then it works" - which is what I do now as well - but have not seen "don't do [... whatever we had done before ...] because that won't work". I don't have the full book right now and not that much time to check closely, but we are kind of approach what that book says after all...

bangerth commented:
Here's another way how you can arrive at the weights: If you call the (naive location for the) midpoint of the four vertices `x0` and the (presumably significantly better) location of the midpoint of the four edge midpoints `x1`, then one could think that extrapolating in this direction would make sense. For example, you could compute `x* = x1 + (x1-x0)/dx * dx` which makes the assumption that you continue to go from `x1` into direction `x1-x0` by one full step `dx`. Then `x*` is exactly the point @kronbichler computes with his weights.

luca-heltai commented:
This is really good! :+1:

kronbichler commented:
I had a closer look at some of the test cases that are failing (I see 190 test failures [1]):

- One challenge is that we many test cases (a few hundreds, so I added that task to #4921) that still use some of the deprecated `Boundary` classes. For `Boundary`, we have the problem that the new way of computing averages on the quads from the whole surrounding and not just the vertices of the quad leads to wrong results (at least for the ones derived from `HyperBallBoundary`). For example, the test `matrix_free/matrix_vector_stokes_noflux.debug` fails with an exception that a quad point is not on the same radius any more, and there are a handful more.

- Then, there are tests like `lac/schur_complement_05` that now fail because some tolerance is not satisfied. I checked that case closely and it appears that the grid was worse before, which gave some spikes in the solution (pressure field) that concentrated refinement to some regions which made it easier to maintain accuracy. But those can be safely adapted I think.

So, there is a lot of work left and I will need a few weeks to fix everything, even if I only start an initial documentation attempt. Unless someone feels a big pressure and wants to help me, I will put a lower priority on it - but it definitely needs to be fixed until the next release.

[1] https://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=12944

tjhei commented:
> some of the deprecated Boundary classes

When do we want to remove Boundary? If we still want to leave it in for 9.0.0 we should also preserve some of the tests. If not, we can rip things out.

Would you mind pulling out your commits about Boundary/Manifold to separate PRs we can merge earlier. Anything to make this PR smaller would be useful.

kronbichler commented:
> When do we want to remove Boundary? If we still want to leave it in for 9.0.0 we should also preserve some of the tests.

I think we keep it in, but deprecate it only. And I agree than some tests should remain, but not hundreds.

drwells commented:
Since a lot of people are still using the boundary classes I would like to leave it in for 9.0.

> For Boundary, we have the problem that the new way of computing averages on the quads from the whole surrounding and not just the vertices of the quad leads to wrong results (at least for the ones derived from HyperBallBoundary).

This sounds like we are going to break user codes if they use `HyperBallBoundary`. Hopefully we can do the same thing that we do in other places and use `dynamic_cast` to switch to the old code path in case a user still has a `Boundary` object; I'll take a look at this later today.

drwells commented:
Some of the test failures are due to different base64 encoding, but a lot have O(1) changes. I'm still working through the failing tests, but here is what I have so far:

- `grid/grid_generator_01a.debug` (error in the first or second significant digit)
- `grid/grid_generator_01b.debug` (error in the first significant digit)
- `grid/grid_hyper_shell.debug` (error in the third significant digit)
- `grid/grid_hyper_shell_02.debug` (error in the first significant digit)
- `grid/grid_hyper_shell_03.debug`
- `grid/grid_out_07.debug` (off by a small amount: the VTK files are identical, but this test tests VTU)
- `grid/grid_output_input.debug`
- `grid/grid_test.debug` (off by the last digit in the output)
- `grid/grid_tools_03.debug`
- `grid/grid_tools_regularize_01.debug`
- `grid/grid_tools_regularize_02.debug`
- `grid/grid_transform.debug`
- `grid/grid_transform_3d.debug`
- `grid/grid_transform_coefficient.debug`
- `grid/manifold_ids_on_boundary_01.debug`
- `grid/project_to_object_01.debug`
- `grid/torus_01.debug`

We do reserve the right to make these kinds of incompatible changes, and it does make the grids a lot better, so I don't mind making this break.

kronbichler commented:
I have done some further analysis regarding the convergence rates and why the new method to compute weights is indeed better. As also remarked by @guidokanschat (off-list), the convergence is ultimately determined by the derivatives of the transformation from unit to real cell, or, more precisely, some of the derivatives of the Jacobian is not O(1) that destroys convergence. To test this hypothesis, I have had a detailed look into the Jacobians for a 2D shell geometry where the spherical/polar manifold is attached to the boundary and we use the usual cells in the interior.

Here is the (0,0) entry of the Jacobian on the initial mesh (6 cells), old to the left, new view to the right:
<img src="https://user-images.githubusercontent.com/8276200/33209783-d968da84-d117-11e7-9d3f-804424fd124e.png" width=330> <img src="https://user-images.githubusercontent.com/8276200/33209789-de7b0c72-d117-11e7-933b-4da6e1b2899d.png" width=330>

Of course, this is a bit unfair since the new approach with the underlying transfinite interpolation is able almost exactly reproduce the circle and of course the Jacobian looks perfectly smooth. But from the distribution of the points inside (I did 27 refinements of each cell in each direction, the mapping is a polynomial of degree 5) one sees why the two are different: Whereas the new approach reproduces a nice blending of the curved boundary, the old (Laplace) weights place points too close together near the inner rim.

It is more interesting to look at differences when the mesh is refined and neither case has continuous Jacobians. Here is the (0,1) of the same case, now refined globally twice, and using an elevation plot to visualize the Jacobians:
<img src="https://user-images.githubusercontent.com/8276200/33210023-d24a155a-d118-11e7-8eec-7ad784052174.png" width=330> <img src="https://user-images.githubusercontent.com/8276200/33210037-d5f82e9e-d118-11e7-8d45-5825b34767b0.png" width=330>

The old way with Laplace smoothing does the weighting by introducing a curvature in the Jacobian entry towards the boundaries. This is unnatural and, as further explained below, the cause for the convergence problems. The new approach does not introduce any additional curvature beyond the curvature introduced by the manifold at the boundary (that is constant within the normal direction to the boundary).

It gets interesting when looking at higher order derivatives. Here is the (0,0,0,0) component of the 2nd derivative of the gradient (3rd derivative of the x-component of the mapping), first on a mesh with `6x2^2=24` cells:
<img src="https://user-images.githubusercontent.com/8276200/33210238-b5d1a702-d119-11e7-9f3f-15aca3567c34.png" width=330> <img src="https://user-images.githubusercontent.com/8276200/33210243-b9ac3086-d119-11e7-8c3b-cbac399e705c.png" width=330>
and now with `6x4^2=96` cells:
<img src="https://user-images.githubusercontent.com/8276200/33210283-dc5809d4-d119-11e7-9ef0-54511b2210e1.png" width=330> <img src="https://user-images.githubusercontent.com/8276200/33210291-e1406360-d119-11e7-8bc6-a838c758bc5f.png" width=330>
When looking at the (small) scale we see that the left decreases as `h^2`, whereas the right one increases as `h^3`. Since the 3rd derivatives have an inherent length scale of `h^3`, the former is `O(h^{-1})` and thus unbounded, whereas the latter is `O(1)` as required by the theory. One can nicely see the effect of the linear blending between the different sides of the geometry that happens within a single cell. Despite involving different cells where the Jacobian is discontinuous because we have different maps, the third derivative is a smooth function again. This is because the (bi-)linear terms of the transfinite interpolation that involve the geometry of the cells are gone at this derivative level and only the derivative of the spherical manifold is left. Note that due to the refinement and application of the manifold only at the boundary, this description is only valid in a rim of one cell, not the whole boundary. But the important point is that the change in the Jacobian is `O(1)` and not `O(1/h)` as in the Laplacian smoothing case.

So this should provide the final verification for the approach we choose by linear interpolation from the surrounding lines into the mid point of the cell in the case of mesh refinement or several points in the case of `MappingQGeneric`.

kronbichler commented:
In case someone wants to replay my experiments, here is the code:
```

#include <deal.II/grid/tria.h>
#include <deal.II/grid/manifold_lib.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/fe/fe_values.h>
#include <deal.II/fe/fe_dgq.h>
#include <deal.II/dofs/dof_handler.h>
#include <deal.II/fe/mapping_q_generic.h>
#include <deal.II/numerics/data_out.h>
#include <deal.II/lac/vector.h>

#include <fstream>

int main()
{
  using namespace dealii;

  const unsigned int dim = 2;
  const unsigned int degree = 5;

  SphericalManifold<dim> spherical;
  MappingQGeneric<dim> mapping(degree);
  Triangulation<dim> tria;
  GridGenerator::hyper_shell(tria, Point<dim>(), 0.5, 1., 6);
  tria.set_all_manifold_ids_on_boundary(1);
  tria.set_manifold(1, spherical);

  for (unsigned int i=0; i<4; ++i)
    {
      FE_DGQ<dim> fe(degree);
      DoFHandler dof_handler(tria);
      dof_handler.distribute_dofs(fe);

      std::vector<Vector<double> > vectors(dim*dim, Vector<double>(dof_handler.n_dofs()));
      std::vector<Vector<double> > vectors_grad(dim*dim*dim, Vector<double>(dof_handler.n_dofs()));
      std::vector<Vector<double> > vectors_grad_grad(dim*dim*dim*dim, Vector<double>(dof_handler.n_dofs()));
      std::vector<Vector<double> > vectors_grad_grad_grad(dim*dim*dim*dim*dim, Vector<double>(dof_handler.n_dofs()));

      FEValues<dim> fe_values(mapping, fe, Quadrature<dim>(fe.get_unit_support_points()),
                              update_jacobians | update_jacobian_grads | update_jacobian_2nd_derivatives | update_jacobian_3rd_derivatives);

      std::vector<types::global_dof_index> dof_indices(fe.dofs_per_cell);
      for (auto cell : dof_handler.active_cell_iterators())
        {
          fe_values.reinit(cell);
          cell->get_dof_indices(dof_indices);
          for (unsigned int d=0; d<dim; ++d)
            for (unsigned int e=0; e<dim; ++e)
              for (unsigned int i=0; i<fe.dofs_per_cell; ++i)
                vectors[d*dim+e](dof_indices[i]) = fe_values.jacobian(i)[d][e];
          for (unsigned int d=0; d<dim; ++d)
            for (unsigned int e=0; e<dim; ++e)
              for (unsigned int f=0; f<dim; ++f)
                for (unsigned int i=0; i<fe.dofs_per_cell; ++i)
                  vectors_grad[(d*dim+e)*dim+f](dof_indices[i]) =
                    fe_values.jacobian_grad(i)[d][e][f];
          for (unsigned int d=0; d<dim; ++d)
            for (unsigned int e=0; e<dim; ++e)
              for (unsigned int f=0; f<dim; ++f)
                for (unsigned int g=0; g<dim; ++g)
                  for (unsigned int i=0; i<fe.dofs_per_cell; ++i)
                    vectors_grad_grad[((d*dim+e)*dim+f)*dim+g](dof_indices[i]) =
                      fe_values.jacobian_2nd_derivative(i)[d][e][f][g];
          for (unsigned int d=0; d<dim; ++d)
            for (unsigned int e=0; e<dim; ++e)
              for (unsigned int f=0; f<dim; ++f)
                for (unsigned int g=0; g<dim; ++g)
                  for (unsigned int h=0; h<dim; ++h)
                  for (unsigned int i=0; i<fe.dofs_per_cell; ++i)
                    vectors_grad_grad_grad[(((d*dim+e)*dim+f)*dim+g)+h](dof_indices[i]) =
                      fe_values.jacobian_3rd_derivative(i)[d][e][f][g][h];
        }

      DataOut<dim, DoFHandler<dim>> data_out;
      data_out.attach_dof_handler(dof_handler);

      for (unsigned int d=0; d<dim; ++d)
        for (unsigned int e=0; e<dim; ++e)
          data_out.add_data_vector(vectors[d*dim+e], "jacobian_" +
                                   std::to_string(d) + std::to_string(e));
      for (unsigned int d=0; d<dim; ++d)
        for (unsigned int e=0; e<dim; ++e)
          for (unsigned int f=0; f<dim; ++f)
            data_out.add_data_vector(vectors_grad[(d*dim+e)*dim+f],
                                     "jacobian_grad_" +
                                     std::to_string(d) +
                                     std::to_string(e) +
                                     std::to_string(f));
      for (unsigned int d=0; d<dim; ++d)
        for (unsigned int e=0; e<dim; ++e)
          for (unsigned int f=0; f<dim; ++f)
            for (unsigned int g=0; g<dim; ++g)
              data_out.add_data_vector(vectors_grad_grad[((d*dim+e)*dim+f)*dim+g],
                                       "jacobian_grad_grad_" +
                                       std::to_string(d) +
                                       std::to_string(e) +
                                       std::to_string(f) +
                                       std::to_string(g));
      for (unsigned int d=0; d<dim; ++d)
        for (unsigned int e=0; e<dim; ++e)
          for (unsigned int f=0; f<dim; ++f)
            for (unsigned int g=0; g<dim; ++g)
              for (unsigned int h=0; h<dim; ++h)
                data_out.add_data_vector(vectors_grad_grad_grad[(((d*dim+e)*dim+f)*dim+g)+h],
                                         "jacobian_grad_grad_grad_" +
                                         std::to_string(d) +
                                         std::to_string(e) +
                                         std::to_string(f) +
                                         std::to_string(g) +
                                         std::to_string(h));

      data_out.build_patches(mapping, 5*degree+2, DataOut<dim>::curved_inner_cells);
      std::ofstream output("grid-" + Utilities::to_string(dof_handler.n_dofs()) + "1.vtu");
      data_out.write_vtu(output);

      tria.refine_global(1);
    }
}
```

kronbichler commented:
I brought down the number of test failures to 11 on tjhei-gcc-serial and around 30 on my machine. I deleted two of the distorted cell tests because they now no longer are distorted. :champagne: 

We are approaching a clean state:
- `numerics/generalized_interpolation_0{2,3,4,5}.debug` and the three matrix-free tests with no-flux constraints will be fixed by #5564.
- `grid/project_to_object_01.debug` should get stable with #5569.
- Finally, the tests `mpi/periodicity_02` and `lac/schur_complement_05` are ill-posed for some reason in the sense that using a more accurate manifold (or more accurate grid, in this case) makes the whole comparison break down. I will try a bit more to see if I can trigger a better refinement by some other means.

Tomorrow I will update the basic documentation and then this PR should be mature for merging. I will need to postpone some of the more in-depth documentation changes because I also want to merge the matrix-free DG code before the holiday seasons and time's running quickly.

guidokanschat commented:
Dear Martin,

Interesting experiments indeed. What’s more important though than continuity of the mapping between cells is the variation inside a cell, since when we do the transformation argument for approximation, we use the minimum and maximum of the Jacobi determinant. The estimate is not sharp, but I think it cannot be improved for higher order shape functions. Question: does max/min blow up for small h?

kronbichler commented:
@guidokanschat No, the min/max of the determinant of the Jacobian does not blow up. It is significantly worse as shown in the following two figures:
<img src="https://user-images.githubusercontent.com/8276200/33546921-a6bdcc8e-d8e2-11e7-801c-48d58f7a0a69.png" width=330> <img src="https://user-images.githubusercontent.com/8276200/33546935-aeb9bd8a-d8e2-11e7-89e2-e1577234ded3.png" width=330>

The elevation in the two plots is the same, so one can clearly see larger values for the old Laplace-style smoothing, in addition to strong variation in the normal direction (which is also visible in the plot for the Jacobian above). So while the Jacobi determinant itself is bounded (changing by a factor of 5 rather than 3 with the optimal parameters), its derivatives are certainly not, at least not if compared to the respective factor of the mesh size depending on the derivative order. So in other words, the Laplace smoothing adds an internal "boundary layer" within the cells, whereas the new weights simply do a linear interpolation between the sides.

kronbichler commented:
I have now finalized all changes and added the documentation. I added the description in the manifold module because that contains the initial discussion of the concepts that are applied here. I removed the WIP label because this should be the final state (except for possible comments that arise now of course).

Three remarks:
- I have tried to use somewhat self-contained commits that group the changes in the tests such that similar changes appear in the same commit. I've done a lot of rebasing, so the remaining 23 commits are still original.
- Review the change in code and documentation. The large number of changes in the tests, touching around 100 tests, is the natural consequence of such a fundamental change. I have verified more than 20 manually with pictures (mostly variants of balls with boundary-attached changes) and looked into changes of the other 80 tests. The details are boring and really boil down to what the figures here are showing.
- This PR must wait for #5564 to get the final 3 matrix-free tests that `tjhei-gcc-serial` complains about fixed.

kronbichler commented:
Here is the current test report on my machine:
```
The following tests FAILED:
	2627 - dofs/dof_renumbering_02.release (Failed)
	4412 - grid/project_to_object_01.release (Failed)
	5349 - lac/schur_complement_05.mpirun=1.debug (Failed)
	5350 - lac/schur_complement_05.mpirun=1.release (Failed)
	5351 - lac/schur_complement_05.mpirun=3.debug (Failed)
	5352 - lac/schur_complement_05.mpirun=3.release (Failed)
	5817 - matrix_free/copy_feevaluation.debug (Failed)
	5983 - matrix_free/matrix_vector_stokes_noflux.debug (Failed)
	5985 - matrix_free/matrix_vector_stokes_notempl.debug (Failed)
	6964 - mpi/renumber_cuthill_mckee_03.mpirun=2.debug (Failed)
	6965 - mpi/renumber_cuthill_mckee_03.mpirun=2.release (Failed)
	8139 - petsc/solver_02.debug (Failed)
	8369 - sharedtria/communicate_active_fe_indices_01.mpirun=3.debug (Failed)
	8370 - sharedtria/communicate_active_fe_indices_01.mpirun=3.release (Failed)
```
Of these, only `project_to_object_01` looks possibly critical after #5575 and #5564 are fixed. The other tests fail due to FMA and roundoff for my machine's configuration.

drwells commented:
@kronbichler Would you please send me the failing output for `project_to_object_01`? Now that #5569 is merged I am not sure what else needs to be fixed for that test.

kronbichler commented:
@drwells I have not looked to deep into it yet but I guess it is the file I modified in #5564, i.e., this one here:
https://github.com/dealii/dealii/pull/5564/files#diff-b79bb94efa40a07427c41a6a36db4693

If it is inconclusive, I can look more into it...

drwells commented:
Given the large change in how we compute things I think that these are okay changes to make to the output.

kronbichler commented:
Rebased - we should now have no test failures on the continuous integration tests.

bangerth commented on doc/doxygen/headers/manifold.h:
```diff
@@ -202,27 +205,20 @@
  *  GridGenerator::hyper_shell (triangulation,
  *                              center, inner_radius, outer_radius,
- *                              4);    // four circumferential cells
+ *                              3);    // four circumferential cells
```
Haven't looked at the rest of the patch, but this at least is funny ;-)

kronbichler replied:
So much about comments and code that should be in phase... 😃

kronbichler commented:
Ping.

kronbichler commented:
Since we all agreed on this change and there were no objections yet, I will merge this tomorrow if no further comments come in. I would like to get the code out now to have as much time and as many users on this as possible before the release.
